lang/src/org/partiql/lang/eval/physical/operators/AggregateOperatorFactory.kt

```diff
@@ -0,0 +1,172 @@
+ *  language governing permissions and limitations under the License.
+ */
+package org.partiql.lang.eval.physical.operators
+
+import com.amazon.ion.IonValue
+import org.partiql.lang.domains.PartiqlPhysical
+import org.partiql.lang.eval.DEFAULT_COMPARATOR
+import org.partiql.lang.eval.ExprValueType
+import org.partiql.lang.eval.MISSING_ANNOTATION
+import org.partiql.lang.eval.physical.EvaluatorState
+import org.partiql.lang.eval.physical.SetVariableFunc
+import org.partiql.lang.eval.relation.relation
+import org.partiql.lang.util.toIntExact
+import java.util.TreeSet
+
+public abstract class AggregateOperatorFactory(name: String) : RelationalOperatorFactory {
+    public final override val key: RelationalOperatorFactoryKey = RelationalOperatorFactoryKey(RelationalOperatorKind.AGGREGATE, name)
+    public abstract fun create(
+        impl: PartiqlPhysical.Impl,
+        source: RelationExpression,
+        keys: List<CompiledGroupKey>,
+        functions: List<CompiledAggregateFunction>
+    ): RelationExpression
+
+    public abstract fun getFunctionFactory(
+        operator: AggregationKind,
+        name: String
+    ): AggregationFunctionFactory
+}
+
+internal class FullAggregateFactory(name: String) : AggregateOperatorFactory(name) {
+
+    private val aggregationFunctionFactories = mapOf(
+        AggregationFunctionFactoryKey(AggregationKind.MAX, "MAX") to MaxFactory("MAX"),
+        AggregationFunctionFactoryKey(AggregationKind.MIN, "MIN") to MinFactory("MIN"),
+        AggregationFunctionFactoryKey(AggregationKind.COUNT, "COUNT") to CountFactory("COUNT"),
+        AggregationFunctionFactoryKey(AggregationKind.SUM, "SUM") to SumFactory("SUM"),
+        AggregationFunctionFactoryKey(AggregationKind.GROUP_AS, "GROUP_AS") to GroupAsFactory("GROUP_AS"),
+        AggregationFunctionFactoryKey(AggregationKind.AVG, "AVG") to AverageFactory("AVG"),
+    )
+
+    override fun getFunctionFactory(operator: AggregationKind, name: String): AggregationFunctionFactory {
+        val key = AggregationFunctionFactoryKey(operator, name)
+        return aggregationFunctionFactories[key] ?: error("Factory for operator ${key.kind} named '${key.name}' does not exist.")
+    }
+
+    override fun create(
+        impl: PartiqlPhysical.Impl,
+        source: RelationExpression,
+        keys: List<CompiledGroupKey>,
+        functions: List<CompiledAggregateFunction>
+    ): RelationExpression =
+        RelationExpression { state ->
+            val sourceIter = source.evaluate(state)
+
+            val aggregationMap = AggregationMap()
+            val seen = List(functions.size) { TreeSet(DEFAULT_COMPARATOR) }
+            val registerSize = state.registers.size
+            val variables = functions.flatMap { function -> function.variables } + keys.map { key -> key.variable }
+            val variablesIndexes = variables.map { variable -> variable.index.value.toIntExact() }.toSet()
+
+            while (sourceIter.nextRow()) {
```

Comment:
I suspect the "double evaluation" and other troubles are caused here.

Consider: 
```
internal class DefaultFooOperator(name: string): FooOperatorFactory(name){
       override fun create(
              impl: PartiqlPhysical.Impl,
             source: RelationExpression
        ) : RelationExpression = {
        // recall that relationExpression. eval -> RelationIterator
                 RelationExpression { state -> 
                       // preparation work
                       val sourceIter = source.evalute(state)
                       val someMaterializedColl = while(sourceIter.nextRow()){
                        // .... do some materialization and preparation work
                       }
                       // the relation iterator produced by operatorFoo, calling eval on operate won't going into the RelationScope defined
                       relation(sourceIter.Type){
                                val iter = someMaterializedColl.iterator()
                                while(iter.hasNext()){
                                       val row = iter.next()
                                       // processing and set up state
                                       yield()
                                }
                       }

                 }
        }
}
```

One observation is that in line 277 of the original code, we call `relationTypeThunk = bexprThunk(env)`. Later, we have another call `val relItr = bexprThunk(env)` (Line 281 in the original Commit). 

For the `foo` operator, it essentially materializes the source twice. I am not sure how to reproduce the exact issue( since the original commit is missing for some reasons?), but I suspect this is the root cause of the issue. 

So one simple fix can be, move the preparation work into the relation scope. 

Let me know if this is helpful.